amqp/src/main/scala/akka/stream/alpakka/amqp/impl/AmqpAsyncFlow.scala

```diff
@@ -0,0 +1,332 @@
+        exitQueue.enqueue(result)
+      } else {
+        log.debug("Pushing {} downstream.", result)
+        push(out, result)
+      }
+    })
+    if (isFinished) closeStage()
+  }
+
+  override def postStop(): Unit = {
+    promise.tryFailure(new RuntimeException("Stage stopped unexpectedly."))
+    super.postStop()
+  }
+
+  override def onFailure(ex: Throwable): Unit = {
+    promise.tryFailure(ex)
+    onFailure(ex)
+  }
+
+  def dequeueAwaitingMessages(tag: DeliveryTag, multiple: Boolean): Iterable[AwaitingMessage[T]]
+
+  def enqueueMessage(tag: DeliveryTag, passThrough: T): Unit
+
+  def messagesAwaitingDelivery: Int
+
+  def noAwaitingMessages: Boolean
+
+  setHandler(
+    in,
+    new InHandler {
+
+      override def onPush(): Unit = {
+        val message: WriteMessage[T] = grab(in)
+        val tag = publish(message)
+
+        scheduleOnce(tag, confirmationTimeout)
+        enqueueMessage(tag, message.passThrough)
+        if (messagesAwaitingDelivery + exitQueue.size < bufferSize && !hasBeenPulled(in)) tryPull(in)
+      }
+
+      override def onUpstreamFailure(ex: Throwable): Unit = {
+        upstreamException = Some(ex)
+        if (isFinished)
+          closeStage()
+        else
+          log.debug("Received upstream failure signal - stage will be failed when all buffered messages are processed")
+
+      }
+
+      override def onUpstreamFinish(): Unit =
+        if (noAwaitingMessages && exitQueue.isEmpty) {
+          promise.success(Done)
+          super.onUpstreamFinish()
+        } else {
+          log.debug("Received upstream finish signal - stage will be closed when all buffered messages are processed")
+        }
+
+      private def publish(message: WriteMessage[T]): DeliveryTag = {
+        val tag: DeliveryTag = channel.getNextPublishSeqNo
+
+        log.debug("Publishing message {} with deliveryTag {}.", message, tag)
+
+        channel.basicPublish(
+          exchange,
+          message.routingKey.getOrElse(routingKey),
+          message.mandatory,
+          message.immediate,
+          message.properties.orNull,
+          message.bytes.toArray
+        )
+
+        tag
+      }
+    }
+  )
+
+  setHandler(
+    out,
+    new OutHandler {
+      override def onPull(): Unit = {
+        if (exitQueue.nonEmpty) {
+          val result = exitQueue.dequeue()
+          log.debug("Pushing enqueued {} downstream.", result)
+          push(out, result)
+        }
+
+        if (isFinished) closeStage()
+        else if (!hasBeenPulled(in)) tryPull(in)
+      }
+    }
+  )
+
+  override protected def onTimer(timerKey: Any): Unit =
+    timerKey match {
+      case tag: DeliveryTag => {
+        log.debug("Received timeout for deliveryTag {}.", tag)
+        onRejection(tag, multiple = false)
+      }
+      case _ => ()
+    }
+
+  private def closeStage(): Unit =
+    upstreamException match {
+      case Some(throwable) =>
+        promise.failure(throwable)
```

Comment:
and here is failure/complete, not tryFailure/tryComplete.

Reply:
`try...` is used just in two cases described in my answert to your comment above. Here I don's see any reason why it wouldn't be safe to call failure/complete.